### PR TITLE
Shapes/contacts

### PR DESCRIPTION
**Pull request by VirginiaBalseiro**
WIP: Defining shapes for Address Book, Groups and Contacts.

# Checklist

- [x] All acceptance criteria are met.
- [ ] Relevant documentation, if any, has been written/updated.
- [ ] The changelog has been updated, if applicable.
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

File: shapes/contacts/group.shex
Comment (megoth) [inline shapes/contacts/group.shex:6]: Should `vcard:hasMember vcard:Individual * OR vcard:Organization * ;`

File: shapes/contacts/group.shex
Comment (megoth) [inline shapes/contacts/group.shex:9]: I think this can be expressed using inverse triple constraints (see page 44 in https://www.slideshare.net/jelabra/shex-by-example-97678930), i.e. 

```
vcard:Group {
  ^vcard:includesGroup vcard:AddressBook ;
}
```

File: shapes/contacts/person.shex
Comment (megoth) [inline shapes/contacts/person.shex:8]: Probably good to check what we can use from https://solidsdk.inrupt.net/sdk/userprofile.shex

File: shapes/README.md
Comment (kevin-inrupt) [inline shapes/README.md:4]: Duplicated "describe" in the same sentence. Maybe "we wanted to explain the"

File: shapes/README.md
Comment (kevin-inrupt) [inline shapes/README.md:22]: "differs" not "differ"

File: shapes/README.md
Comment (kevin-inrupt) [inline shapes/README.md:19]: ShEx shape also includes:
`  acl:owner             IRI +`

File: shapes/README.md
Comment (kevin-inrupt) [inline shapes/README.md:32]: See previous comment about address-book.shex shape also containing `acl:owner`.

File: shapes/README.md
Comment (kevin-inrupt) [inline shapes/README.md:35]: `contacts/group-index.shex`

File: shapes/README.md
Comment (kevin-inrupt) [inline shapes/README.md:51]: Text needs to be revised to explain the group.shex shape.

File: shapes/README.md
Comment (kevin-inrupt) [inline shapes/README.md:1]: There seems to be a large number of revisions needed in the README based on the updates to the various ShEx shapes. Instead of rewriting in the PR, probably worth reworking and resubmitting, even in a separate PR after the ShEx shapes are approved.

File: shapes/contacts/person.shex
Comment (kevin-inrupt) [inline shapes/contacts/person.shex:36]: While there is nothing wrong with the Email description, it is very limited only having an email address. It would be nice to have at least the type of email. Plus it would be good to have the regex to validate the email format. The https://solidsdk.inrupt.net/sdk/userprofile.shex describes an email as:
```
  <#UserProfileEmail> {
    a [vcard:Dom vcard:Home vcard:ISDN vcard:Internet vcard:Intl vcard:Label vcard:Parcel vcard:Postal vcard:Pref vcard:Work vcard:X400] {1}
      // rdfs:label "Type"@en
      // rdfs:label "Tipo"@es ;
    vcard:value xsd:string /^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$/ {1}
      // rdfs:label "Email"@en 
      // rdfs:label "Correo"@es
      // :prefix "mailto:";
  }
```
A few comments:

- Do not have to include `rdfs:label`s, although good for translations.
- ShEx does not have the concept of dynamically populated lists, hence the hard-coded list values.

File: shapes/contacts/person.shex
Comment (kevin-inrupt) [inline shapes/contacts/person.shex:28]: Would like to include Address type similar to email and phone types.

While not explicitly needed, may be worth putting min and max string lengths in the shape to stop users entering novels in the fields.

File: shapes/contacts/person.shex
Comment (kevin-inrupt) [inline shapes/contacts/person.shex:40]: Similar comments as for email - type, regex, string lengths.

Refer to https://solidsdk.inrupt.net/sdk/userprofile.shex:

```<#UserProfileTelephone> {
    a [vcard:Dom vcard:Home vcard:ISDN vcard:Internet vcard:Intl vcard:Label vcard:Parcel vcard:Postal vcard:Pref vcard:Work vcard:X400] {1}
      // rdfs:label "Type"@en
      // rdfs:label "Tipo"@es ;
    vcard:value xsd:string /^\+?[0-9]+[0-9-]*[0-9]$/ {1}
      // rdfs:label "Phone"@en 
      // rdfs:label "TelÃ©fono"@es
      // :prefix "tel:" ;
    }
```

**Review by kevin-inrupt — CHANGES_REQUESTED** [review]
Comments on README and ShEx shapes. The ShEx shapes comments are just suggestions, but would provide more functionality and validation.

File: shapes/contacts/person.shex
Comment (VirginiaBalseiro) [reply shapes/contacts/person.shex:36]: > [vcard:Dom vcard:Home vcard:ISDN vcard:Internet vcard:Intl vcard:Label vcard:Parcel vcard:Postal vcard:Pref vcard:Work vcard:X400]

Definitely makes sense to have email type. One question: do we need all those types? Other than Home and Work, they all seem to have been deprecated.

File: shapes/contacts/person.shex
Comment (VirginiaBalseiro) [reply shapes/contacts/person.shex:28]: Another question: what do you think would be a good upper limit? The string length limit would be for:

- Street name
- Locality
- Region
- Country name

Some street names and city names can be very long in some parts of the world: https://en.wikipedia.org/wiki/List_of_long_place_names

File: shapes/README.md
Comment (VirginiaBalseiro) [reply shapes/README.md:1]: I applied most of the changes suggested but I agree, we should revise the whole README once we have the final shapes.

Comment (megoth) [reply shapes/README.md:1]: I did some changes, but now waiting for @timbl final review, then we'll see what needs to be changed further.

File: shapes/contacts/person.shex
Comment (megoth) [reply shapes/contacts/person.shex:37]: @VirginiaBalseiro I think we went through the Regexes with @timbl ?

File: shapes/contacts/person.shex
Comment (megoth) [reply shapes/contacts/person.shex:28]: @timbl did not want min and max string lengths in the shape - he argued it should be specific to the the form models, not the shape itself.

Comment (megoth) [reply shapes/contacts/person.shex:28]: What do you mean with `Would like to include Address type similar to email and phone types.` btw? I think @VirginiaBalseiro has addressed that?

Comment (VirginiaBalseiro) [reply shapes/contacts/person.shex:28]: I did address the types, this comment was still unresolved due to the max length question but the comment is prior to my latest changes where I added types to Address, email and Tel.

File: shapes/contacts/person.shex
Comment (VirginiaBalseiro) [reply shapes/contacts/person.shex:37]: @megoth I don't remember going over the regexes, please feel free to remove them if you think we shouldn't have them there.

Comment (megoth) [reply shapes/contacts/person.shex:37]: Ok, I probably remember wrong then.

@timbl Do you think we should use the Regexes suggested in the ShEx-shapes, or those in the SHACL shape, or something different?

File: shapes/contacts/person.shex
Comment (megoth) [reply shapes/contacts/person.shex:28]: Since Tim said he didn't want the MAXLENGTH and MINLENGTH, I'll resolve this thread.

Comment (kevin-inrupt) [reply shapes/contacts/person.shex:28]: We may want to give this some more thought. Originally, the React SDK v1 did not use the min and max lengths. However, when the team implemented the ShEx Shape -> Form Model  Converter, the min and max lengths (along with other field parameters) were _very_ useful to help select the Form Model field type to use for each of the ShEx shape fields. For example, based on the string max length, the Converter could decide whether to use Textboxes or Text Areas. Similar with Integer fields to decide whether it should use a pre-populated dropdown (a few integers), a slider, or Textbox.

File: shapes/contacts/person.shex
Comment (kevin-inrupt) [reply shapes/contacts/person.shex:37]: Re: Email Types. A smaller selection of email (and phone) tyoes would be fine. The list shown are the ones used in Data Browser. If the types could be dynamically generated (and in the User Profile Form Model used by React SDK v1 Generator https://solidsdk.inrupt.net/sdk/userprofile.ttl), the list of types would be:
- Email: Home, Work
- Phone: Cell, Fax, Pager, Text, TextPhone, Video, Voice

File: shapes/contacts/person.shex
Comment (kevin-inrupt) [reply shapes/contacts/person.shex:28]: Re: Address Type. If @VirginiaBalseiro has added one, then great.

Comment (kevin-inrupt) [reply shapes/contacts/person.shex:28]: One thing to consider with the Address is the fact there are only one Street Name and one Locality. In way of an example, in England, locals like naming their houses and having multiple locales. When I was growing up, my official address was:

- The Bungalow, (Name of house)
- River Road, 
- West Walton (Village)
- Wisbech (Town)
- Cambs (County)
- PE14 7EX (Post Code)
- UK

I have similar issues with family and friends, and I am sure this is not unique to the UK.

If we want to be sophisticated, we could enhance the shape to allow for these kinds of addresses.

**Review by ajacksified — APPROVED** [review]
We can and should merge this, and continue the discussion about where we should publicly host our shapes.

File: shapes/contacts/person.shex
Comment (kevin-inrupt) [inline shapes/contacts/person.shex:43]: For compatibility with Data Browser, Phone types should be Cell, Fax, Pager, Text, TextPhone, Video, Voice

**Review by kevin-inrupt — CHANGES_REQUESTED** [review]
One small change request on Phone types.

File: shapes/contacts/person.shex
Comment (megoth) [reply shapes/contacts/person.shex]: Should be resolved now ^_^

File: shapes/contacts/person.shex
Comment (megoth) [reply shapes/contacts/person.shex:28]: We could change it to `vcard:street-address xsd:string + ;` which would allow for one or more street addresses? (And same for `vcard:locality xsd:string +;`.)

Comment (ajacksified) [reply shapes/contacts/person.shex:28]: We should use whatever DB uses for now and continue the conversation about what's optimal in a separate ticket.

**Review by kevin-inrupt — APPROVED** [review]
OK with applied changes.

**Pull request merged**
